### ISSUE TITLE
Create better official image Dockerfile

### ISSUE THREAD

**Issue by stevvooe** [docker/distribution#582]
Many users have run into issues with running the official image. The main issue is that the configuration file is the same one provided for development. This would entail the following:

1. Move the official build to a separate project. This allows a fixed interface to the Dockerfile, along with a README covering usage, such as correct storage driver configuration for the official image.
2. Such a Dockerfile would be built such that only the resulting binary and dependencies are including. The resulting image would be smaller and lighter weight.
3. A default configuration file that is easily overridden for common use cases. Right now, this file is just the "Development" configuration file, which includes unnecessary and underspecified items.

**Comment by amouat** [comment]
:+1: 

Also, putting the config file at an easier location than `/go/src/github.com/docker/distribution/cmd/registry/config.yml` would be appreciated.

**Comment by stevvooe** [comment]
@amouat It was originally in `/etc/docker/registry/config.yml` for this exact reason. It was changed in 1211d49. Not sure why. @bfirsh 

We'll cover this with item 3 above.

**Comment by amouat** [comment]
Great thanks.

**Comment by kenjones-cisco** [comment]
:+1:

~550 MB is quite large given the registry part is only about ~31 MB.

**Comment by aaronlehmann** [comment]
The only way I know of to do (2) is to have two separate Dockerfiles; one which does the build and another that copies the binaries and support files into a fresh container. This would be scripted in the build system so that a "release" container can be built with a make target.

The notes here about CA certificate dependencies may be relevant: https://blog.codeship.com/building-minimal-docker-containers-for-go-applications/.

I'm not sure I understand what (1) entails.

**Comment by dmp42** [comment]
@aaronlehmann have a look at what swarm does:

 * https://github.com/docker/swarm
 * https://github.com/docker/swarm-library-image

**Comment by aaronlehmann** [comment]
Ah, I had looked at swarm, but didn't know about swarm-library-image. Thanks, that's helpful. I'll make something similar for distribution (and notary). Who should I contact about creating a new GitHub repo for the library image?

**Comment by aaronlehmann** [comment]
I've made progress creating a minimal image based on swarm-library-image. But I don't think we can push the minimalism as far as swarm does.

swarm compiles a static binary and sticks it in an empty container with a few support files. The registry uses cgo for rados, so we end up with a dynamically linked executable. It's theoretically possible to build static binaries that use cgo, but it introduces a bunch of problems (i.e.: `/usr/src/go/src/os/user/lookup_unix.go:75: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking`). Also, trying to link statically to librados fails with a lot of C++ linkage issues. It may be possible to work around these, but I haven't put in the effort so far.

Bottom line, I think it's best for us to stick with a dynamically linked executable, and build on top of an image that has the C/C++ standard libraries, and the ability to easily install librados (so probably ubuntu?).

Or we could disable the rados driver in the official image.

**Comment by aaronlehmann** [comment]
PR for review here: https://github.com/docker/distribution-library-image/pull/1

**Comment by dmp42** [comment]
@gierschv can you chip in on this?

**Comment by gierschv** [comment]
@dmp42 Yes, I replied in the PR, thanks for the ping.

**Comment by RichardScothern** [comment]
Closing as complete.
